optuna/storages/_journal/file.py

```diff
@@ -0,0 +1,175 @@
+from optuna.storages._journal.base import BaseJournalLogStorage
+
+
+LOCK_FILE_SUFFIX = ".lock"
+RENAME_FILE_SUFFIX = ".rename"
+
+
+class JournalFileBaseLock(abc.ABC):
+    @abc.abstractmethod
+    def acquire(self) -> bool:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def release(self) -> None:
+        raise NotImplementedError
+
+
+@experimental_class("3.1.0")
+class JournalFileSymlinkLock(JournalFileBaseLock):
+    """Lock class for synchronizing processes for NFSv2 or later.
+
+    On acquiring the lock, link system call is called to create an exclusive file. The file is
+    deleted when the lock is released. In NFS environments prior to NFSv3, use this instead of
+    :class:`~optuna.storages.JournalFileOpenLock`
+
+    Args:
+        filepath:
+            The path of the file whose race condition must be protected.
+    """
+
+    def __init__(self, filepath: str) -> None:
+        self._lock_target_file = filepath
+        self._lockfile = filepath + LOCK_FILE_SUFFIX
+        self._lockrenamefile = self._lockfile + str(uuid.uuid4()) + RENAME_FILE_SUFFIX
+
+    def acquire(self) -> bool:
+        while True:
+            try:
+                os.symlink(self._lock_target_file, self._lockfile)
+                return True
+            except OSError as err:
+                if err.errno == errno.EEXIST:
+                    continue
+                else:
+                    raise err
+            except BaseException:
+                self.release()
+                raise
+
+    def release(self) -> None:
+        try:
+            os.rename(self._lockfile, self._lockrenamefile)
+            os.unlink(self._lockrenamefile)
+        except OSError:
+            raise RuntimeError("Error: did not possess lock")
+        except BaseException:
+            os.unlink(self._lockrenamefile)
+            raise
+
+
+@experimental_class("3.1.0")
+class JournalFileOpenLock(JournalFileBaseLock):
+    """Lock class for synchronizing processes for NFSv3 or later.
+
+    On acquiring the lock, open system call is called with the O_EXCL option to create an exclusive
+    file. The file is deleted when the lock is released. This class is only supported when using
+    NFSv3 or later on kernel 2.6 or later. In prior NFS environments, use
+    :class:`~optuna.storages.JournalFileSymlinkLock`.
+
+    Args:
+        filepath:
+            The path of the file whose race condition must be protected.
+    """
+
+    def __init__(self, filepath: str) -> None:
+        self._lockfile = filepath + LOCK_FILE_SUFFIX
+        self._lockrenamefile = self._lockfile + str(uuid.uuid4()) + RENAME_FILE_SUFFIX
+
+    def acquire(self) -> bool:
+        while True:
+            try:
+                open_flags = os.O_CREAT | os.O_EXCL | os.O_WRONLY
+                os.close(os.open(self._lockfile, open_flags))
+                return True
+            except OSError as err:
+                if err.errno == errno.EEXIST:
+                    continue
```

Comment:
Ditto

Reply:
Fixed in https://github.com/optuna/optuna/pull/3854/commits/3917a29e3796673f3c5cdbc4e7fa7eb0b1aa7f4f.